app/scripts/controllers/backup.js

```diff
@@ -0,0 +1,57 @@
+import { exportAsFile, prependZero } from '../../../ui/helpers/utils/util';
+
+export default class BackupController {
+  constructor(opts = {}) {
+    const {
+      preferencesController,
+      addressBookController,
+      trackMetaMetricsEvent,
+    } = opts;
+
+    this.preferencesController = preferencesController;
+    this.addressBookController = addressBookController;
+    this._trackMetaMetricsEvent = trackMetaMetricsEvent;
+  }
+
+  async restoreUserData(jsonString) {
+    const { preferences, addressBook } = JSON.parse(jsonString);
+    preferences && this.preferencesController.store.updateState(preferences);
+    addressBook && this.addressBookController.update(addressBook, true);
+
+    this._trackMetaMetricsEvent({
+      event: 'User restored data from backup',
+      category: 'backup',
+    });
+  }
+
+  async backupUserData() {
+    const userData = {
+      preferences: this.preferencesController.store.getState(),
+      addressBook: this.addressBookController.state,
+    };
+
+    const result = JSON.stringify(userData);
+
+    const date = new Date();
+
+    const prefixZero = (num) => prependZero(num, 2);
+
+    /*
+     * userData.YYYY_MM_DD_HH_mm_SS e.g userData.2022_01_13_13_45_56
+     * */
+    const userDataFileName = `userData.${date.getFullYear()}_${prefixZero(
+      date.getMonth() + 1,
+    )}_${prefixZero(date.getDay())}_${prefixZero(date.getHours())}_${prefixZero(
+      date.getMinutes(),
+    )}_${prefixZero(date.getDay())}.json`;
+
+    exportAsFile(userDataFileName, result);
```

Comment:
We should move the `exportAsFile` to a utility file in the `shared/` directory

Reply:
Alternatively, this could be called from the UI. This utility uses the DOM, so it won't work in the MV3 service worker anyway